File: weaverlib/src/main/java/com/antourage/weaverlib/other/networking/ApiClient.java
package com.antourage.weaverlib.other.networking;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiClient {

    public static String BASE_URL = "";
    public static final String HEADER_TOKEN = "Authorization";
    private static final String HEADER_LANGUAGE = "Accept-Language";

    private static WebService webService;
    private static Retrofit retrofit;
    private static ApiClient apiClient;

    private ApiClient() {
        OkHttpClient client = defaultClient();

        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL + "api/v1/")
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(new LiveDataCallAdapterFactory())
                .build();
        webService = retrofit.create(WebService.class);
    }

    public static ApiClient getClient() {
        if (apiClient == null || !retrofit.baseUrl().toString().equals(BASE_URL + "api/v1/")) {
            apiClient = new ApiClient();
        }
        return apiClient;
    }

    public static Retrofit getRetrofitInstance() {
        return retrofit;
    }

    public static String getTokenForRequest() {
        String token = "";
//        if (App.Companion.getPrefsHelper() != null){
//            token = App.Companion.getPrefsHelper().getToken();
//            if (token == null) token = "";
//        }
        return token;
    }

    public WebService getWebService() {
        return webService;
    }

    private OkHttpClient defaultClient() {
        return new OkHttpClient.Builder().addInterceptor(chain -> {
            Request request = chain.request().newBuilder()
                    .addHeader(HEADER_TOKEN, getTokenForRequest())
                    .addHeader(HEADER_LANGUAGE, "en")
                    .build();
            return chain.proceed(request);
        }).build();
    }
}

File: ant-viewver/src/main/java/com/antourage/weaverlib/other/networking/ApiClient.kt
package com.antourage.weaverlib.other.networking

import com.antourage.weaverlib.ConfigManager.BASE_URL
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

internal object ApiClient {
    private const val VERSION_SUFFIX = "api/v3/"

    lateinit var apiService: ApiService
    private var retrofit: Retrofit? = null
    private var httpClient: OkHttpClient? = null

    fun getWebClient(): ApiClient {
        if (retrofit == null || (retrofit?.baseUrl().toString() != BASE_URL + VERSION_SUFFIX)) {
            buildRetrofit()
        }
        return this
    }

    fun getHttpClient(): OkHttpClient {
        if (httpClient == null) {
            httpClient = buildOkHttpClient()
        }
        return httpClient as OkHttpClient
    }

    //region Private

    private fun buildRetrofit() {
        httpClient = buildOkHttpClient()
        retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL + VERSION_SUFFIX)
            .client(httpClient!!)
            .addConverterFactory(GsonConverterFactory.create())
            .addCallAdapterFactory(LiveDataCallAdapterFactory())
            .build()
        apiService = retrofit?.create(ApiService::class.java)!!
    }

    private fun buildOkHttpClient(): OkHttpClient {
        //TODO delete before release
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY)

        val builder = OkHttpClient.Builder()
        builder
            .addInterceptor(loggingInterceptor)
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(15, TimeUnit.SECONDS)
        return builder.build()
    }

    //endregion
}